File: utils/property_checks.py
"""Property Checks for numbers and strings."""
import math


_SQUARE_BIT_PATTERNS = frozenset((0x00, 0x01, 0x04, 0x09, 0x10, 0x11,
                                  0x19, 0x21, 0x24, 0x29, 0x31, 0x39))


def is_perfect_square(n):
    """
    http://stackoverflow.com/q/295579
    """
    if n < 0:
        return False

    x = n & 0x3F
    if x not in _SQUARE_BIT_PATTERNS:
        return False

    sqrt = int(math.sqrt(n))

    return sqrt*sqrt == n


def is_permutation(a, b):
    return is_permutation_str(str(a), str(b))


def is_permutation_str(a, b):
    return sorted(a) == sorted(b)


def is_triangular_number(n):
    return is_perfect_square(8*n + 1)


def is_pentagonal_number(n):
    x = (((24*n + 1) ** 0.5) + 1) / 6
    return x == int(x)


def is_hexagonal_number(n):
    x = (((8*n + 1) ** 0.5) + 1) / 4
    return x == int(x)


def is_heptagonal_number(n):
    x = (((40*n + 9) ** 0.5) + 3) / 10
    return x == int(x)


def is_octagonal_number(n):
    x = (((3*n + 1) ** 0.5) + 1) / 3
    return x == int(x)


def is_prime(n):
    if n <= 3:
        return n >= 2
    if n % 2 == 0 or n % 3 == 0:
        return False
    for i in range(5, int(n ** 0.5) + 1, 6):
        if n % i == 0 or n % (i + 2) == 0:
            return False
    return True


def is_palindrome(n):
    str_n = str(n)
    return str_n == str_n[::-1]


def is_binary_palindrome(n):
    str_n = bin(n)[2:].lstrip('0')
    return str_n == str_n[::-1]


def _is_pandigital_closure():
    PANDIGITAL = ['1', '2', '3', '4', '5', '6', '7', '8', '9']

    def is_pandigital(s, n):
        """
        Return False or None if not a palindrome true otherwise.

        None is returned if the number given has more digits than N.
        This allows anyone incrementing a counter to check that s will now
        never match and allow a loop break while still maintaining a false
        boolness if this isnt needed.
        """
        len_s = len(s)
        if len_s > n:
            return None
        elif len_s != n:
            return False

        for p in PANDIGITAL[:n]:
            if p not in s:
                return False

        return True
    return is_pandigital

is_pandigital = _is_pandigital_closure()

File: go/src/utils/property_checks.go
package utils

import (
	"math"
	"sort"
	"strconv"
)

var squareBitPatterns = map[int]bool{
	0x00: true, 0x39: true, 0x24: true, 0x09: true,
	0x19: true, 0x01: true, 0x29: true, 0x10: true,
	0x21: true, 0x04: true, 0x31: true, 0x11: true,
}

func SqrtInt(n int) int {
	return int(math.Sqrt(float64(n)))
}

func IsPerfectSquare(n int) bool {
	if n < 0 {
		return false
	}
	x := n & 0x0F
	// Uses Go's default map fill behaviour here,
	// Anything not in BitPatterns will give a false
	// as a result so no need to do a two item map check.
	if !squareBitPatterns[x] {
		return false
	}

	sq := SqrtInt(n)
	return sq*sq == n
}

func IsTriangular(n int) bool {
	return IsPerfectSquare(8*n + 1)
}

func IsPentagonal(n int) bool {
	x := (math.Sqrt(float64(24*n+1)) + 1) / 6
	return x == math.Floor(x)
}

func IsHexagonal(n int) bool {
	x := (math.Sqrt(float64(8*n+1)) + 1) / 4
	return x == math.Floor(x)
}

func IsHeptagonal(n int) bool {
	x := (math.Sqrt(float64(40*n+9)) + 3) / 10
	return x == math.Floor(x)
}

func IsOctagonal(n int) bool {
	x := (math.Sqrt(float64(3*n+1)) + 1) / 3
	return x == math.Floor(x)
}

func IsPrime(n int) bool {
	if n <= 3 {
		return n >= 2
	}
	if ModInt(n, 2) == 0 || ModInt(n, 3) == 0 {
		return false
	}
	for i := 5; i < int(math.Sqrt(float64(n)))+1; i += 6 {
		if ModInt(n, i) == 0 || ModInt(n, i+2) == 0 {
			return false
		}
	}
	return true
}

func ModInt(a, b int) int {
	l := 0
	r := a
	for l < r {
		m := (l + r) / 2
		if (a - m*b) >= b {
			l = m + 1
		} else {
			r = m
		}
	}

	return a - l*b
}

func IsPalindrome(s string) bool {
	n := len(s)
	r := []rune(s)

	for i := 0; i < n/2; i++ {
		if r[i] != r[n-i-1] {
			return false
		}
	}
	return true
}

type runeSlice []rune

func (p runeSlice) Len() int           { return len(p) }
func (p runeSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
func (p runeSlice) Less(i, j int) bool { return p[i] < p[j] }

func IsPermutation(a, b string) bool {
	if len(a) != len(b) {
		return false
	}
	r1 := runeSlice(a)
	r2 := runeSlice(b)
	sort.Sort(r1)
	sort.Sort(r2)
	for k, v := range r1 {
		if v != r2[k] {
			return false
		}
	}
	return true
}

func IsPandigital(n int) bool {
	r := runeSlice(strconv.Itoa(n))
	sort.Sort(r)
	start := 0
	if r[0] == '0' {
		if len(r) != 10 {
			return false
		}
		start = 48
	} else if r[0] == '1' {
		start = 49
	} else {
		return false
	}
	for k, v := range r {
		// Does not support 0 in pandigitals
		if v != rune(start+k) {
			return false
		}
	}
	return true
}
